File: 1170-shortest-common-supersequence/1170-shortest-common-supersequence.cpp
class Solution {
public:
    int LCS(int i, int j, string &str1, string &str2, vector<vector<int>> &dp){
        if(i == 0 || j == 0) return 0;

        if(dp[i][j] != -1) return dp[i][j];

        if(str1[i-1] == str2[j-1]){
            return dp[i][j] = 1 + LCS(i-1, j-1, str1, str2, dp);
        }

        return dp[i][j] = max(LCS(i-1, j, str1, str2, dp), LCS(i, j-1, str1, str2, dp));
    }

    string shortestCommonSupersequence(string str1, string str2) {
        int n = str1.size();
        int m = str2.size();
        vector<vector<int>> dp(n+1, vector<int>(m+1, -1));
        int lcs = LCS(n, m, str1, str2, dp);

        int i = n;
        int j = m;
        string ans = "";

        while(i >0 && j >0){
            if(str1[i-1] == str2[j-1]){
                ans += str1[i-1];
                i--;
                j--;
            }else if(dp[i-1][j] > dp[i][j-1]){
                ans += str1[i-1];
                i--;
            }else{
                ans += str2[j-1];
                j--;
            }
        }

        while(i > 0){
            ans += str1[i-1];
            i--;
        }

        while(j >0 ){
            ans += str2[j-1];
            j--;
        }

        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: 1170-shortest-common-supersequence/1170-shortest-common-supersequence.py
class Solution:
    def LCS(self, i, j, str1, str2, dp):
        if i == 0 or j == 0:
            return 0

        if dp[i][j] != -1:
            return dp[i][j]

        if str1[i-1] == str2[j-1]:
            dp[i][j] = 1 + self.LCS(i-1, j-1, str1, str2, dp)
            return dp[i][j]
        
        dp[i][j] = max(self.LCS(i-1, j, str1, str2, dp), self.LCS(i, j-1, str1, str2, dp))
        
        return dp[i][j]

    def shortestCommonSupersequence(self, str1: str, str2: str) -> str:
        n = len(str1)
        m = len(str2)
        dp = [[-1 for _ in range(m+1)] for _ in range(n+1)]

        dummy = self.LCS(n,m,str1, str2,dp)

        i = n
        j = m

        ans = ""

        while i > 0 and j > 0:
            if str1[i-1] == str2[j-1]:
                ans += str1[i-1]
                i-=1
                j-=1
            elif dp[i-1][j] > dp[i][j-1]:
                ans += str1[i-1]
                i-=1
            else:
                ans += str2[j-1]
                j-=1
        
        while i > 0:
            ans += str1[i-1]
            i-=1

        while j > 0:
            ans += str2[j-1]
            j-=1

        return ans[::-1]


